### PR TITLE
Convert Program TS files to JSON. Update script to generate JSON from spreadsheet data.

### PR DESCRIPTION
**Pull request by matelau**
Convert Program TS files to JSON. Update script to generate JSON from spreadsheet data.
This is done to make the data files more consistent which will aid us in how we integrate incentive admin data modifications back into the API. See: The [Mini Brief](https://docs.google.com/document/d/1nM-4Utwbe68tcY60Rw94a9HZQp1l5PWobkQdSEqUQfI/edit?usp=sharing) for more details. 

## Links

- [_Asana_](https://app.asana.com/0/1207304265252553/1207827064980175/f)
- [Incentive Admin Mini-brief](https://docs.google.com/document/d/1nM-4Utwbe68tcY60Rw94a9HZQp1l5PWobkQdSEqUQfI/edit?usp=sharing)

## Changes

-  Convert state_programs.ts files to JSON
- update scripts/lib/authority-and-program-updater.ts to generate JSON files when executing 
- update README

## Test Plan

- run `yarn ts-node scripts/generate-misc-state-data.ts [TWO CHAR STATE CODE]` and verify corresponding `state/programs.json` file is created. 
- run `yarn dev` 
- hit the api with a request for a SLU incentive ie: `http://localhost:3000/api/v1/calculator?zip=89015&include_beta_states=true&owner_status=homeowner&household_income=2000&tax_filing=single&household_size=1` 

## Next steps

- Update Import and Export workflow in Incentive Admin.

### CODE REVIEW

File: scripts/lib/authority-and-program-updater.ts [hunk 1]
```diff
@@ -386,20 +305,19 @@ export class AuthorityAndProgramUpdater {
...
-    const filePath = PROGRAMS_DIR + `${this.state.toLowerCase()}_programs.ts`;
+    const filePath =
+      PROGRAMS_DIR + '/' + this.state.toUpperCase() + '/programs.json';
     const formatOptions = await prettier.resolveConfig(filePath);
     if (!formatOptions) {
       throw new Error(`Could not retrieve Prettier config for ${filePath}`);
     }
     formatOptions.filepath = filePath;
-    const tsFileContent = await createProgramsContent(
+    const JSONFileContent = await createProgramsContent(
```
Comment (veekas) [inline scripts/lib/authority-and-program-updater.ts:315]: nittiest nit to have ever nitted: I tend to prefer lowercasing all acronyms that are at the beginning of a camelCase variable name as we do with [`nyJson`](https://github.com/rewiringamerica/api.rewiringamerica.org/blob/asa/convert-programs-to-json/scripts/ny-empower-income-limits.ts#L49-L50) and [`csvContent`](https://github.com/rewiringamerica/api.rewiringamerica.org/blob/asa/convert-programs-to-json/scripts/ny-empower-income-limits.ts#L49-L50)

File: test/snapshots/v1-dc-20303-state-city-lowincome.json [hunk 2]
```diff
@@ -132,7 +132,7 @@
       ],
       "authority_type": "city",
       "authority": "dc-dc-sustainable-energy-utility",
-      "program": "Solar for All Single- Family Program",
+      "program": "Solar for All Single-Family Program",
```
Comment (veekas) [inline test/snapshots/v1-dc-20303-state-city-lowincome.json:135]: ... I thought i fixed this back in March 🤪

**Review by veekas — APPROVED** [review]
does anyone remember why we used TS files in the first place? this is much cleaner

**Comment by matelau** [comment]
> does anyone remember why we used TS files in the first place? this is much cleaner

I'm not sure, maybe @oyamauchi knows?

File: scripts/lib/authority-and-program-updater.ts (hunk 1 above)
Comment (matelau) [reply scripts/lib/authority-and-program-updater.ts]: updated!

**Review by oyamauchi — CHANGES_REQUESTED** [review]
The purpose was to validate that each incentive had a valid `program` value using JSON Schema validation, rather than doing so in explicit test code. This made sense back when there were, like, ten programs, then grew absurd as the number of programs went into the hundreds.

In that vein, though, could you add a test that does that check?

File: test/data/schemas.test.ts [hunk 3]
```diff
@@ -233,11 +233,23 @@ const isURLValid = (url: string): boolean => {
...
+        tap.ok(isURLValid(url), `${programId} has invalid URL`);
+      } else {
+        tap.ok(false, `${programId} has invalid URL`);
+      }
     }
   }
 });
 
+test('program JSON files match schemas', async tap => {
+  const ajv = new Ajv();
```
Comment (matelau) [inline test/data/schemas.test.ts:246]: @oyamauchi I added this test that tests all programs using the program schema. Let me know if you think this should be done on a state by state basis instead, but this matches how the programs are loaded by the application.

File: test/snapshots/v1-dc-20303-state-city-lowincome.json (hunk 2 above)
Comment (veekas) [reply test/snapshots/v1-dc-20303-state-city-lowincome.json:135]: oh yeah, I probably only updated it there. mystery solved!

File: src/data/programs.ts [hunk 4]
```diff
@@ -1,14 +1,43 @@
...
 
-export type Program = {
-  name: LocalizableString;
-  url: LocalizableString;
-};
+export const PROGRAM_SCHEMA = {
+  title: 'Program',
+  type: 'object',
+  patternProperties: {
+    '^[a-zA-Z]': {
```
[the remark this answers is not in the document]
Comment (matelau) [reply src/data/programs.ts]: updated to use the following pattern ` '^.*$'`

File: src/data/programs.ts [hunk 5]
```diff
@@ -1,14 +1,43 @@
...
+          },
+          required: ['en'],
+        },
+        url: {
+          type: 'object',
+          properties: {
+            en: { type: 'string' },
+            es: { type: 'string' },
+          },
+          required: ['en'],
```
[the remark this answers is not in the document]
Comment (matelau) [reply src/data/programs.ts]: This seems to work for ajv, but wasn't working well for generating the typescript type. So, I ended up separating the type def from the json schema, looks like we did something similar with ira incentives.   
image

File: test/data/schemas.test.ts [hunk 6]
```diff
@@ -233,11 +233,31 @@ const isURLValid = (url: string): boolean => {
 test('all programs have valid URLs', async tap => {
   for (const [programId, data] of Object.entries(PROGRAMS)) {
     for (const url of Object.values(data.url ?? [])) {
-      tap.ok(isURLValid(url), `${programId} has invalid URL`);
+      if (typeof url === 'string') {
```
[the remark this answers is not in the document]
Comment (matelau) [reply test/data/schemas.test.ts]: This is related to the type generated with `FromSchema` not being  adequately defined. see: https://github.com/rewiringamerica/api.rewiringamerica.org/pull/536#discussion_r1686925556

**Pull request merged**